gef-ui/server/controllers/companies-house/index.test.js

```diff
@@ -168,11 +178,61 @@ describe('controllers/about-exporter', () => {
       });
     });
 
-    it('redirects user to `problem with service` page if there is an issue with any of the api', async () => {
+    it(`redirects to the 'application details' page when the API call returns company data and the status query is set to 'change'`, async () => {
+      when(getCompanyByRegistrationNumberMock)
+        .calledWith({ registrationNumber: MOCK_COMPANY_REGISTRATION_NUMBERS.VALID, userToken })
+        .mockResolvedValue({ company: {} });
+
+      mockRequest.query.status = 'change';
+      mockRequest.body.regNumber = MOCK_COMPANY_REGISTRATION_NUMBERS.VALID;
+
+      await validateCompaniesHouse(mockRequest, mockResponse);
+
+      expect(mockResponse.redirect).toHaveBeenCalledWith('/gef/application-details/123');
+    });
+
+    it('rerenders the page with the correct errors the API call returns an error object', async () => {
+      const invalidCompanyRegistrationNumberErrorObject = {
+        errRef: 'regNumber',
+        errMsg: 'Enter a valid Companies House registration number.',
+      };
+      const mappedErrorObject = {
+        someErrors: 'some errors',
+      };
+
+      when(getCompanyByRegistrationNumberMock)
+        .calledWith({ registrationNumber: MOCK_COMPANY_REGISTRATION_NUMBERS.INVALID_TOO_SHORT, userToken })
+        .mockResolvedValue(invalidCompanyRegistrationNumberErrorObject);
+      when(validationErrorHandler).calledWith([invalidCompanyRegistrationNumberErrorObject]).mockReturnValue(mappedErrorObject);
+
+      mockRequest.body.regNumber = MOCK_COMPANY_REGISTRATION_NUMBERS.INVALID_TOO_SHORT;
+
+      await validateCompaniesHouse(mockRequest, mockResponse);
+
+      expect(mockResponse.render).toHaveBeenCalledWith('partials/companies-house.njk', {
+        errors: mappedErrorObject,
+        regNumber: MOCK_COMPANY_REGISTRATION_NUMBERS.INVALID_TOO_SHORT,
+        dealId: '123',
+        status: undefined,
+      });
+    });
+
+    it(`redirects to the 'problem with service' page if any of the API calls throw errors`, async () => {
       const mockedRejection = { response: { status: 400, message: 'Whoops' } };
       api.getApplication.mockRejectedValueOnce(mockedRejection);
+
       await validateCompaniesHouse(mockRequest, mockResponse);
+
       expect(mockResponse.render).toHaveBeenCalledWith('partials/problem-with-service.njk');
     });
+
+    it(`logs the error if any of the API calls throw errors`, async () => {
```

Comment:
Can we add one for `500` as well please.

Reply:
Added.